_posts/en/pages/2016-01-18-segwit-wallet-dev.md

```diff
@@ -0,0 +1,124 @@
+
+<pre><code>[nVersion][marker][flag][txins][txouts][witness][nLockTime]</code></pre>
+
+Format of <code>nVersion</code>, <code>txins</code>, <code>txouts</code>, and <code>nLockTime</code> are same as traditional serialization.
+
+The <code>marker</code> MUST be <code>0x00</code>.
+
+The <code>flag</code> MUST be a 1-byte non-zero value. Currently, <code>0x01</code> MUST be used.
+
+The <code>witness</code> is a serialization of all witness data of the transaction. Each txin is associated with a witness field. A witness field starts with a var_int to indicate the number of stack items for the txin. It is followed by stack items, with each item starts with a var_int to indicate the length. Witness data is NOT script and is not restricted by the 520-byte push limit.
+
+A non-witness program txin MUST be associated with an empty witness field, represented by a <code>0x00</code>. If all txins are not witness program, a transaction MUST be serialized with the traditional format (exception: coinbase transaction).
+
+### Transaction ID
+
+Each transaction will have 2 IDs.
+
+Definition of <code>txid</code> remains unchanged: the double SHA256 of the traditional serialization.
+
+A new <code>wtxid</code> is defined: the double SHA256 of the new serialization with witness data. If a transaction is transmitted without any witness data, its <code>wtxid</code> is same as the <code>txid</code>.
+
+The txid remains the primary identifier of a transaction. Particularly, it is used in the txin when referring to a previous output.
+
+### Standard script types
+
+#### Pay-to-Public-Key-Hash (P2PKH)
+P2PKH is the most common <code>scriptPubKey</code> template defined by Satoshi Nakamoto, allows simple payment to a single public key. The format is:
+
+<pre><code>scriptPubKey (25 bytes): OP_DUP OP_HASH160 < 20-byte-pubkey-hash > OP_EQUALVERIFY OP_CHECKSIG</code></pre>
+
+To spend a P2PKH output, the <code>scriptSig</code> is
+
+<pre><code>scriptSig: < sig > < pubkey ></code></pre>
+
+where the <code>RIPEMD160(SHA256(pubkey))</code> is equal to the <code>20-byte-pubkey-hash</code> in scriptPubKey.
+
+#### Pay-to-Script-Hash (P2SH)
+P2SH is defined in BIP16. It allows payment to arbitrarily complex scripts with a fixed length <code>scriptPubKey</code>. The format is:
+
+<pre><code>scriptPubKey (23 bytes): OP_HASH160 <20-byte-script-hash> OP_EQUAL</code></pre>
+
+To spend a P2SH output, the <code>scriptSig</code> is
+
+<pre><code>scriptSig: <...> <...> <...> < redeemScript ></code></pre>
+
+where the <code>RIPEMD160(SHA256(redeemScript))</code> is equal to the <code>20-byte-script-hash</code> in <code>scriptPubKey</code>. The <code>redeemScript</code> is deserialized and evaluated with the remaining data in the <code>scriptSig</code>.
+
+#### Pay-to-Witness-Public-Key-Hash (P2WPKH)
+P2WPKH is a new defined in BIP141. Similar to P2PKH, it allows simple payment to a single public key. The format is:
```

Comment:
grammar here: "is a new defined in BIP141", needs to be "P2WPKH is a new script type defined in BIP141"